homeassistant/components/camera/__init__.py

```diff
@@ -0,0 +1,273 @@
+"""
+homeassistant.components.camera
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Component to interface with various cameras.
+
+The following features are supported:
+-Recording
+-Snapshot
+-Motion Detection Recording(for supported cameras)
+-Automatic Configuration(for supported cameras)
+-Creation of child entities for supported functions
+-Collating motion event images passed via FTP into time based events
+-Returning recorded camera images and streams
+-Proxying image requests via HA for external access
+-Converting a still image url into a live video stream
+-A service for calling camera functions
+
+Upcoming features
+-Camera movement(panning)
+-Zoom
+-Light/Nightvision toggling
+-Support for more devices
+-A demo entity
+-Expanded documentation
+"""
+import requests
+from requests.auth import HTTPBasicAuth
+import logging
+import time
+import re
+from homeassistant.helpers.entity import Entity
+from homeassistant.const import (
+    ATTR_ENTITY_PICTURE,
+    HTTP_NOT_FOUND,
+    ATTR_ENTITY_ID,
+    )
+
+from homeassistant.helpers.entity_component import EntityComponent
+
+
+DOMAIN = 'camera'
+DEPENDENCIES = ['http']
+GROUP_NAME_ALL_CAMERAS = 'all_cameras'
+SCAN_INTERVAL = 30
+ENTITY_ID_FORMAT = DOMAIN + '.{}'
+
+SWITCH_ACTION_RECORD = 'record'
+SWITCH_ACTION_SNAPSHOT = 'snapshot'
+
+SERVICE_CAMERA = 'camera_service'
+
+STATE_RECORDING = 'recording'
+
+DEFAULT_RECORDING_SECONDS = 30
+
+# Maps discovered services to their platforms
+DISCOVERY_PLATFORMS = {}
+
+FILE_DATETIME_FORMAT = '%Y-%m-%d_%H-%M-%S-%f'
+DIR_DATETIME_FORMAT = '%Y-%m-%d_%H-%M-%S'
+
+REC_DIR_PREFIX = 'recording-'
+REC_IMG_PREFIX = 'recording_image-'
+
+STATE_STREAMING = 'streaming'
+STATE_IDLE = 'idle'
+
+
+# pylint: disable=too-many-branches
+def setup(hass, config):
+    """ Track states and offer events for sensors. """
+
+    component = EntityComponent(
+        logging.getLogger(__name__), DOMAIN, hass, SCAN_INTERVAL,
+        DISCOVERY_PLATFORMS)
+
+    component.setup(config)
+
+    # -------------------------------------------------------------------------
+    # CAMERA COMPONENT ENDPOINTS
+    # -------------------------------------------------------------------------
+    # The following defines the endpoints for serving images from the camera
+    # via the HA http server.  This is means that you can access images from
+    # your camera outside of your LAN without the need for port forwards etc.
+
+    # Because the authentication header can't be added in image requests these
+    # endpoints are secured with session based security.
+
+    # pylint: disable=unused-argument
+    def _proxy_camera_image(handler, path_match, data):
+        """ Proxies the camera image via the HA server. """
+        entity_id = path_match.group(ATTR_ENTITY_ID)
+
+        camera = None
+        if entity_id in component.entities.keys():
+            camera = component.entities[entity_id]
+
+        if camera:
+            response = camera.get_camera_image()
```

Comment:
Should we add support for a mimetype to the camera entity? 

Will make it possible to:
 - set the correct header
 - disable a mjpeg stream if they do not support jpeg